### PR TITLE
Make errorprone happy and silent

### PR DESCRIPTION
**Pull request by joaoandremartins**
- Makes errorprone happy and silent,
- Several Java8 compliant changes.

### CODE REVIEW

**Comment by patflynn** [comment]
it looks like you need to rebase.

**Comment by joaoandremartins** [comment]
Seems like you and I spent yesterday working on the same things.

**Comment by patflynn** [comment]
@joaoandremartins are you watching the repo?  You should have gotten notifications of the PRs yesterday.

**Comment by patflynn** [comment]
are you using a tool to generate these refactorings or doing them manually?

**Comment by joaoandremartins** [comment]
I wasn't, but I will now.

The main motivation for this PR was getting rid of all the errorprone warnings, including the futures one and that enum mutable field. That required changes to a few files, and I did the rest of the refactors manually. I doubt there is a tool that can perform most of the refactors automatically, since they required changing the logic a little bit.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardFacetEditor.java [hunk 1]
```diff
@@ -166,50 +167,42 @@ private boolean isManagedLibrariesEnabled() {
...
-          ModuleRootManager manager = ModuleRootManager.getInstance(context.getModule());
-          ModifiableRootModel model = manager.getModifiableModel();
-
-          model.addLibraryEntry(addedLibrary);
-
-          for (OrderEntry orderEntry : model.getOrderEntries()) {
-            if (orderEntry.getPresentableName().equals(addedLibrary.getName())) {
-              ((ExportableOrderEntry) orderEntry).setScope(scope);
-            }
+      AppEngineStandardMavenLibrary.getLibraryByMavenDisplayName(addedLibrary.getName()).ifPresent(
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardFacetEditor.java]: Can you explain a bit why you added this check?

**Comment by etanshaul** [comment]
this may be an annoying request, but since these changes are easily submit separately, could you split them up? perhaps one pr to update the library stuff, another to change to lambdas, another to change to Optionals.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardFacetEditor.java (hunk 1 above)
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardFacetEditor.java]: Specifically I guess - how could it be that the added library is not already an AppEngineStandardMavenLibrary?

Comment (joaoandremartins) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardFacetEditor.java]: This is just the translation of the previous code. Notice that you're checking if library is null.

I don't think it will ever happen that library isn't present. But, since getLibraryByMavenDisplayName() returns an Optional, I think this is the best way to work with that.

Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardFacetEditor.java]: Oh yea - I missed that part because of the refactoring.

Could you restore it to the format I had it to prevent such deep nesting? I think it makes it easier to read.

**Comment by joaoandremartins** [comment]
I thought the way we were going about this was to make Java8 changes whenever we came across the files.

It would be a lot of work to split this into separate PRs, and I don't think it's worth the trouble. All the code here, besides replacing RepositoryLibraryProperties in the enum, should just be a direct translation of Java 7 - Java 8. I didn't add/remove anything.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardFacetEditor.java (hunk 1 above)
Comment (joaoandremartins) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardFacetEditor.java]: I try to avoid using isPresent() has hard as I can, since it throws away the benefit of using Optionals. I prefer my approach, it's more functional style, and I think it's more readable too.

Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardFacetEditor.java]: fair enough. i do like the use of ifPresent, but the nesting is really deep. I can live with it. you could extract the lambda inside of it to another method too.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/application/AppEngineApplicationCreateDialog.java [hunk 2]
```diff
@@ -147,8 +147,8 @@ private void trackApplicationCreateFailure() {
         .ping();
   }
 
-  private void setStatusMessageAsync(final String message, final boolean isError) {
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/application/AppEngineApplicationCreateDialog.java]: does errorProne really complain about this? even though its always true (since we don't have other use cases yet), it feels weird to be hardcoding it to true in this method. Not a big deal though

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineFacetConfiguration.java [hunk 3]
```diff
@@ -69,19 +71,11 @@ public void writeExternal(Element element) throws WriteExternalException {
...
-      if (mavenLibrary != null) {
-        configuredLibraries.add(mavenLibrary);
-      }
-    }
-
-    return configuredLibraries;
+    return Arrays.stream(
+        LibraryTablesRegistrar.getInstance().getLibraryTable(project).getLibraries())
+        .map(library ->
+            AppEngineStandardMavenLibrary.getLibraryByMavenDisplayName(library.getName()).get())
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineFacetConfiguration.java]: is this null safe? what about only adding it to the set if its non null, does this still do that?

Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineFacetConfiguration.java]: the null safe part of my question is referring to the direct get() you are doing on the Optional.

**Comment by patflynn** [comment]
@joaoandremartins  
Keep separate concerns in separate PRs whenever possible.  

You should have broken this PR up as:
 PR #1: Fix unused future warnings
 PR #2: Fix mutable enum
 PR #3: Refactor to Java8 closures

For PR #3, we should discuss our approach to Java 8 migration.  I haven't personally come to a conclusion on what the right approach is here, and I'm open to suggestions.  

I had taken the attitude that flyby formatting changes and copyright updates was OK, but I know they annoyed @etanshaul  recently and I see his point.  Probably something worth discussing.. possibly in the retrospective this afternoon?

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardLibraryPanel.java [hunk 4]
```diff
@@ -65,67 +60,47 @@ public AppEngineStandardLibraryPanel(boolean enabled) {
...
+      JCheckBox checkbox = (JCheckBox) event.getItem();
+
+      if (!checkbox.isSelected()) {
+        objectifyCheckBox.setSelected(false);
+        endpointsCheckBox.setSelected(false);
       }
     });
-    objectifyCheckBox.addItemListener(new AppEngineApiDependencyCheckboxListener());
-    endpointsCheckBox.addItemListener(new AppEngineApiDependencyCheckboxListener());
+    objectifyCheckBox.addItemListener(this::listenAppEngineApiDependencyCheckbox);
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardLibraryPanel.java]: I like the refactoring, but not the name choice for the method. Maybe handleAppEngineApiDependencyCheckboxEvent?

**Comment by joaoandremartins** [comment]
Yes, I understand where you guys are coming from. However, as I mentioned before, all these changes are direct translations to Java 8 idioms, without any actual behavioural changes, except for removing the mutable field from the enum.

I feel that your suggested changes highly affect productivity, and I meant to approach this problem some time in conversation.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardLibraryPanel.java [hunk 5]
```diff
@@ -65,67 +60,47 @@ public AppEngineStandardLibraryPanel(boolean enabled) {
...
-                    ? AppEngineStandardMavenLibrary
-                    .getLibraryByDisplayName(libraryCheckbox.getText())
-                    : null;
-              }
-            }),
-        Predicates.notNull()
-    ));
+    return libraries.stream()
+        .filter(JCheckBox::isSelected)
+        .map(
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardLibraryPanel.java]: what about the filtering out of the nulls?

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardLibraryPanel.java [hunk 6]
```diff
@@ -65,67 +60,47 @@ public AppEngineStandardLibraryPanel(boolean enabled) {
...
+            library ->
+                AppEngineStandardMavenLibrary.getLibraryByDisplayName(library.getText()).get())
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardLibraryPanel.java]: again, isn't this not-null safe to do a get() on an Optional like this?

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardLibraryPanel.java [hunk 7]
```diff
@@ -65,67 +60,47 @@ public AppEngineStandardLibraryPanel(boolean enabled) {
...
-    for (JCheckBox libraryCheckbox : libraries) {
-      libraryCheckbox.setSelected(availableLibraryNames.contains(libraryCheckbox.getText()));
-    }
+    Set<String> availableLibraryNames = mavenLibraries.stream()
+        .map(AppEngineStandardMavenLibrary::getDisplayName)
+        .collect(toSet());
+
+    libraries.stream()
+        .filter(library -> availableLibraryNames.contains(library.getText()))
+        .forEach(checkbox -> checkbox.setSelected(true));
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardLibraryPanel.java]: I don't think this is the same behavior. Previously the logic was also unselecting libraries. Now, it looks like you are only ticking them.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineFacetConfiguration.java (hunk 3 above)
Comment (chanseokoh) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineFacetConfiguration.java]: > does this still do that?

I don't think it will. `map()` in any functional language never removes an element, i.e., the resulting collection has the same size.

Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineFacetConfiguration.java]: @chanseokoh - agreed, I have a bad habit of phrasing my observations as questions :)

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardFacetEditor.java (hunk 1 above)
Comment (patflynn) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardFacetEditor.java]: I imagine that this is going to take some time to get used to, but for now I'm finding this style to be a significant step back in readability.

Comment (patflynn) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardFacetEditor.java]: @elharo any thoughts on this?  This seems related to our discussion this morning.

Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardFacetEditor.java]: my opinion on this is that I like the style in general but a couple things could improve readability:

1) the nesting due to the lambda plus the write action etc. is hurting readability. If the contents of the ifPresent lamdba were extracted out to logically named methods I think it would help greatly.
2) Extract out AppEngineStandardMavenLibrary.getLibraryByMavenDisplayName(addedLibrary.getName()) to a named variable.

Comment (chanseokoh) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardFacetEditor.java]: > If the contents of the ifPresent lamdba were extracted out to logically named methods

If it becomes a method, I'm guessing just passing the method name would work too (http://stackoverflow.com/a/24228366/1701388).

Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardFacetEditor.java]: true, but I didn't mean to imply that it necessarily should be split that way into 1 method

Comment (joaoandremartins) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardFacetEditor.java]: I think factoring out the lambda contents into a new method would actually decrease readability, since that lambda essentially contains the bulk of logic of the method. i.e., `afterLibraryAdded()` would be a one-liner with a method reference, which voids the utility of the reference method.

We could get rid of the ifPresent() by making sure that getLibraryByMavenDisplayName() can always return something, though I'm not familiar enough with the code to say if that can be done.

Finally, I think the following is actually pretty readable, but feel free to disagree.

screenshot from 2017-02-22 15 25 46

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/application/AppEngineApplicationCreateDialog.java (hunk 2 above)
Comment (joaoandremartins) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/application/AppEngineApplicationCreateDialog.java]: errorprone doesn't, but the IJ linter does, so I thought I'd remove it. I was wondering if this method had been called before actually using that argument, but then that call was deleted. If there aren't any plans to start using this with a variable isError, then I'd like to delete based on YAGNI.

`setStatusMessage()` is called with variable isError, which is why I didn't cut it there as well.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardFacetEditor.java (hunk 1 above)
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardFacetEditor.java]: its ok.

I think my suggestion (2) would help a bit.

Also (and this was there before any of your changes), you could update the for loop in the WriteAction to be java8'y

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineFacetConfiguration.java (hunk 3 above)
Comment (joaoandremartins) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineFacetConfiguration.java]: What I think happens is map maps to Optional.empty(), in case of empty optionals returned by the mapping method (i.e., no nulls involved), and then later, the collector skips empty optionals.

The example in http://www.oracle.com/technetwork/articles/java/java8-optional-2175753.html seems to indicate that, and I think the Java 8 in Action book covers that, too.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardLibraryPanel.java (hunk 4 above)
Comment (joaoandremartins) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardLibraryPanel.java]: Done.

**Comment by patflynn** [comment]
The fact that there are multiple points of debate about behavior preservation leads me to believe that this PR skews to the wrong side of risk/cost/benefit. 

Even if it were worth it the PR needs to be broken up, with the benefit that the safe and non-controversial changes will be approved very quickly.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineFacetConfiguration.java (hunk 3 above)
Comment (chanseokoh) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineFacetConfiguration.java]: Below gives me `[Optional.empty, Optional.empty, Optional.empty]`.

```java
  public static void main(String[] args) {
    String[] arr = new String[]{ null, null, null };
    System.out.println(Arrays.stream(arr).map(e -> Optional.empty()).collect(Collectors.toList()));
  }
```

Comment (joaoandremartins) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineFacetConfiguration.java]: Yes. Turns out that the recommended way appears to be:

libraries.stream()
        .filter(JCheckBox::isSelected)
        .map(
            library ->
                AppEngineStandardMavenLibrary.getLibraryByDisplayName(library.getText()))
        .filter(Optional::isPresent)
        .map(Optional::get)
        .collect(toSet());

The Optional creator discusses it in http://stackoverflow.com/questions/22725537/using-java-8s-optional-with-streamflatmap. Java 9 will enable Optional.stream(), which makes flatMap here work as intended.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardFacetEditor.java (hunk 1 above)
Comment (joaoandremartins) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardFacetEditor.java]: So, if I understand (2) correctly, I'd extract the whole method to a separate method, and then just call it from afterLibraryAdded()? If so, why not just call the whole thing there?

Yes, I like getting rid of the for loop. Also, to get rid of that WriteAction boilerplate, I'm considering using `ApplicationManager.getApplication().runWriteAction()`;. Should be equivalent?

Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardFacetEditor.java]: For 2 thats not what I meant. I just meant something simply like this for readability:

```
AppEngineStandardMavenLibrary libary = 
  AppEngineStandardMavenLibrary.getLibraryByMavenDisplayName(addedLibrary.getName());

library.ifPresent(.. => ..
```

**Comment by patflynn** [comment]
@joaoandremartins if you still want to make these changes let's move the conversation to multiple PRs addressing the separate concerns that are contained in this one.

I just want to reiterate that this PR should not be merged.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardLibraryPanel.java (hunk 7 above)
Comment (joaoandremartins) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineStandardLibraryPanel.java]: Ah, good catch!

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineSupportProvider.java [hunk 8]
```diff
@@ -211,42 +212,30 @@ static void removeMavenLibraries(final Set<AppEngineStandardMavenLibrary> librar
...
-              final Library library = libraryTable.getLibraryByName(displayName);
-              if (library != null) {
-                libraryTable.removeLibrary(library);
-
-                for (OrderEntry orderEntry : model.getOrderEntries()) {
-                  if (orderEntry.getPresentableName().equals(library.getName())) {
-                    model.removeOrderEntry(orderEntry);
-                  }
+    ApplicationManager.getApplication().invokeLater(() -> {
+      new WriteAction() {
```
Comment (joaoandremartins) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineSupportProvider.java]: I wonder if we could just use a single thread here. We're running invokeLater, which then runs writeAction. I wonder would simply running ApplicationManager.getApplication().runWriteAction() cut it here?

Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/AppEngineSupportProvider.java]: I don't know the answer off hand. runWriteAction does not invoke the Runnable asynchronously so it may be a behavior change. Tt would need to be thoroughly tested.

Regardless, along the lines of Patrick's comment, this change alone could be its own PR.

**Comment by etanshaul** [comment]
You may be on it already, but I'm seeing files unrelated to this PR. I don't know why these merge issues keep happening assuming you are doing an actual merge, not a rebase? Maybe we can compare workflows since I never have to do this (I merge directly from IJ and do not rebase).

**Comment by joaoandremartins** [comment]
Java 8 changes removed. PTAL.

@etanshaul that would be most useful! I usually run git rebase origin/master from the command line. This is what happens when I attempt a Push -> Merge from IntelliJ. It has never worked for me, in the half dozen of times I've tried it.

**Comment by patflynn** [comment]
LGTM on the enum refactor. @etanshaul owns this code so I'll let him be the approver.
